docs/guides/testing-keycloak.rst

```diff
@@ -0,0 +1,111 @@
+Testing Keycloak integration
+############################
+
+Keycloak integration can be tested using the related methods provided in
+:class:`sssd_test_framework.roles.keycloak.Keycloak`.
+
+Multhost test configuration requirements
+----------------------------------------
+
+Testing with Keycloak integration requires additional systems configured to
+provide a Keycloak server and the ipa-tuura SCIMv2 **bridge** service:
+
+.. code-block:: yaml
+
+    domains:
+    - id: sssd
+      hosts:
+      - hostname: idm.ipa.test
+        role: ipa
+        ssh:
+          host: HOSTNAME_IPA
+          username: root
+          password: Secret123
+        config:
+          client:
+          ipa_domain: ipa.test
+
+      - hostname: bridge.ipa.test
+        role: client
+        ssh:
+          host: HOSTNAME_BRIDGE
+          username: root
+          password: Secret123
+
+      - hostname: keycloak.ipa.test
+        role: keycloak
+        ssh:
+          host: HOSTNAME_KEYCLOAK
+          username: root
+          password: Secret123
+        config:
+          keycloak:
+            keycloak_server: keycloak.ipa.test
+            keycloak_admin_user: admin
+            keycloak_admin_password: Secret123
+
+It should be noted that at this time, the ipa-tuura bridge service does not
+yet have its own role.  Instead it is defined as a client for remote access if
+necessary and is managed via the Keycloak SCIM storage plugin.
+
+Adding a Keycloak Integration Domain
+------------------------------------
+
+Keycloak integration requires the Keycloak server to add an Integration Domain
+that will point to the identity provider being tested (IPA, AD, LDAP, etc).
+This is accomplished using the method
+:meth:`sssd_test_framework.roles.keycloak.KeycloakDomain.add`.   After Adding
+the Integration Domain, you can add users to keycloak and they should be synced
+to the identity provider.  At that point, you should be able to use them like
+any user created for that provider.
+
+Test example
+============
+
+.. code-block:: python
+
+    from __future__ import annotations
+
+    import pytest
+
+    from pytest_mh import Topology, TopologyDomain
+
+    from sssd_test_framework.roles.client import Client
+    from sssd_test_framework.roles.keycloak import Keycloak
+    from sssd_test_framework.roles.ipa import IPA
+    from sssd_test_framework.roles.generic import GenericProvider
+
+
+    @pytest.mark.topology(
+        "ipa",
+        Topology(TopologyDomain("sssd", client=1, keycloak=1, ipa=1)),
+        keycloak="sssd.keycloak[0]", client="sssd.client[0]", ipa="sssd.ipa[0]", provider="sssd.ipa[0]"
```

Comment:
provider can be omitted, I think, since this is tightly bound to IPA, therefore you can not use topology parametrization. Or can you?

Reply:
Yes, it could be but, I have some questions/issues with using existing ones for the tests.   

I didn't think I could use the existing topologies because the tests don't need an NFS instance.  Could I just define the nfs server in the mhc.yaml as the client or ipa server and ignore it in the tests?

And if I used existing KnownTopologies, how would I define the keycloak role fixture to be used with the test?

```
@pytest.mark.topology(KnownTopology.LDAP)
@pytest.mark.topology(KnownTopology.AD)
@pytest.mark.topology(KnownTopology.IPA)
def test_example(keycloak: Keycloak, client: Client, provider: GenericProvider):
```

Or should I create new topologies for each like:
```
Keycloak_IPA = (keycloak, client, ipa, provider)
Keycloak_AD = (keycloak, client, ad, provider)
Keycloak_LDAP = (keycloak, client, ldap, provider)
```

Reply:
So the bridge is running SCIMv2+SSSD, with SSSD connected to the backends, right? The second example is what you want. But the client should be ipa_tuura_bridge or scimclient or something more specific and Keycloak should probably also switch to IPA_TUURA, however IPA_TUURA_AD looks weird and IPA is not involved at all here, right?

Other way that would be maybe better is to define these topologies in ipa-tuura tests only (i.e. within your test code) and then you can use just IPA, AD, Samba, LDAP.  Because you have sligthly different needs then what SSSD has. Then you can freely use client.

Reply:
Yes, SSSD connects to the backends in the ipa-tuura setup.   Also, correct that there is no IPA involved in the ipa-tuura setup with AD.

I see the logic in something like IPA_TUURA_{PROVIDER} name.  Especially considering that ipa-tuura is what is actually connected to the backend provider. (via SSSD).  Part of the reason I was thinking Keycloak_{PROVIDER} was that the goal of using ipa-tuura is to connect Keycloak to the backend providers.   So I was thinking End to End Keycloak to AD, Keycloak to IPA, Keycloak to LDAP.

I could also just define those with the ipa-tuura tests but, how should I document this?  Or should I create a generic, simple Keycloak toplogy that just containers Keycloak as the provider (also will be useful later I think)?

Reply:
So the idea is not to have everything under sssd-test-framework, that's why it is all separated into pytest-mh and sssd-test-framework. pytest-mh can be used by any projects, sssd-test-framework is for sssd project but most idm projects can benefit from it. But extensions specific to particular project should go to the project, not here.

The way I see it, Keycloak role can go here, because SSSD can benefit it when we will include tests for external IdPs. But these topologies containing the bridge are specific to ipa-tuura and should go to this project - you can add it either to the same repository as tests or in another repository and you can use it together with sssd-test-framework.

Therefore you will be able to have custom roles and topologies that makes sense for ipa-tuura but not for sssd and you can provide shorter and meaningful names. Documentation should go to ipa-tuura project.

It's up to you, I won't deny the contribution of course. If you choose to place it all here, IPA_TUURA_(PROVIDER) is in my opinion the way to go. But yes, the idea is to share as much as possible, but sssd-test-framework is not supposed to be the ultimate do-it-all repository, because you will reach these kind of decisions.

Reply:
Ok, I'll put the ipa-tuura specific topologies with those tests and create a generic keycloak one that we can use later for external IdP.

Reply:
I moved the ipa-tuura specific topologies as mentioned above.